src/autorest.bicep/src/type-generator.ts

```diff
@@ -19,31 +20,83 @@ export function generateTypes(host: AutorestExtensionHost, definition: ProviderD
     host.message({ Channel: Channel.Information, Text: message, });
   }
 
-  function processResourceBody(fullyQualifiedType: string, definition: ResourceDefinition) {
-    const { descriptor, putRequest, putParameters, putSchema, getSchema, } = definition;
-    const nameSchemaResult = parseNameSchema(
-      putRequest,
-      putParameters,
-      schema => parseType(schema, schema),
-      (name) => factory.addType(new StringLiteralType(name)));
-
-    if (!nameSchemaResult.success) {
-      logWarning(`Skipping resource type ${fullyQualifiedType} under path '${putRequest.path}': ${nameSchemaResult.error}`);
-      return
+  function getResourcePath(definition: ResourceDefinition) {
+    return (definition.putOperation ?? definition.getOperation)?.request.path;
+  }
+
+  function getNameType(fullyQualifiedType: string, definition: ResourceDefinition) {
+    function getSchema(op: ResourceOperationDefintion) {
+      const r = getNameSchema(op.request, op.parameters);
+
+      if (!r.success) {
+        logWarning(`Skipping resource type ${fullyQualifiedType} under path '${op.request.path}': ${r.error}`);
+        return
+      }
+
+      return r.value;
     }
 
-    if (!nameSchemaResult.value) {
-      logWarning(`Skipping resource type ${fullyQualifiedType} under path '${putRequest.path}': failed to obtain a name value`);
+    const {putOperation, getOperation} = definition;
+    const nameLiterals = new Set<string>();
+    const nameTypes = new Set<BuiltInTypeKind>();
+    for (const ns of [putOperation ? getSchema(putOperation) : undefined, getOperation ? getSchema(getOperation) : undefined]) {
+      if (!ns) {
+        continue;
+      }
+
+      if (ns.type === 'parameterized') {
+        const {schema} = ns;
+        if (schema instanceof ConstantSchema && toBuiltInTypeKind(schema.valueType) === BuiltInTypeKind.String) {
+          nameLiterals.add(schema.value.value);
+        } else if (schema instanceof ChoiceSchema || schema instanceof SealedChoiceSchema) {
+          const enumValues = getValuesForEnum(schema);
+          if (enumValues.success) {
+            const {values, closed} = enumValues.value;
+            values.forEach(v => nameLiterals.add(v));
+            if (!closed) {
+              nameTypes.add(BuiltInTypeKind.String);
+            }
+          }
+        } else {
+          nameTypes.add(toBuiltInTypeKind(schema));
+        }
+      } else {
+        nameLiterals.add(ns.value);
+      }
+    }
+
```

Comment:
First off, I apologize for the general lack of comments in this repo. The code is already difficult to follow, and with the handling of more corner cases, it's becoming more complex. No need to address in this PR, but going forwards I think it might be worth thinking about starting to add some comments to explain some of the more complex logic.